Buffered shared-memory integrals for the Green-MBPT code

gauravharsha opened the pull request:
This pull request overcomes the slow step of reading chunks of integrals one at a time. This involves the following changes:
1. Move the handling of integrals to [GreenPhys/green-integrals](https://github.com/green-phys/green-integrals), where shared memory implementation is introduced.
2. Modify the usage of integral objects in GW and GF2 kernels.

gauravharsha commented:
The tests do not fail. In this PR, I marked one of the tests to be skipped for the time being. On Perlmutter, I get the following output from CTest:
```shell
test cases: 1 | 1 skipped
assertions: 8 | 8 passed

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   5.11 sec

The following tests FAILED:
	 40 - MBPT Solver (Failed)
```
The skipped test is considered as a failure.

iskakoff commented on CMakeLists.txt:
```diff
@@ -11,6 +11,7 @@ endif ()
 set(GREEN_RELEASE origin/main)
 
 set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_FLAGS "-O3 -g -fno-omit-frame-pointer")
```
what? no, why do you enforce `-O3`?

gauravharsha replied:
Removed `-O3`

iskakoff commented on CMakeLists.txt:
```diff
@@ -34,6 +35,12 @@ if (BLAS_FOUND)
     endif ()
     add_compile_definitions(EIGEN_USE_BLAS)
 endif ()
+find_package(Eigen3 REQUIRED)
```
All these dependencies are not needed as they are obtained through dependencies

gauravharsha replied:
Removed all dependencies from the parent library - dependencies take care of them wherever needed.

iskakoff commented on geintegrals_test.cpp:
Test should be in `test` directory, hard coded paths should not be used.

gauravharsha replied:
On second thoughts, this is not a good unit test. Will remove it totally

iskakoff replied:
Great!

iskakoff commented on mpi_read_lock.cpp:
What is this file needed for?

gauravharsha replied:
Removed.

iskakoff commented on src/green/mbpt/common_defs.h:
```diff
@@ -31,7 +31,6 @@
 
 namespace green::mbpt {
 
-  static const std::string INPUT_VERSION = "0.2.4";
```
Why do you want to remove it? `input.h5` has nothing to do with `green-integrals` repository, and we check `input.h5` version as well. More rational would be to introduce separate version check in `green-integrals`.

gauravharsha replied:
I agree, we should keep the checks within green-mbpt. But input and integrals are generated in the same step, so we should have just one centralized test where both the input file and integrals file is checked. Otherwise there will be extra book-keeping and possibilities for bugs in future updates.

iskakoff replied:
Two main things in support of moving integrals version check and corresponding tests away from `green-mbpt`.
 
 - There are different coulomb integrals that mbpt can use (now it's two and hopefully it will be more) and each of them technically could have their own version.
 - What if someone introduce a bug into version checking in `green-integrals`. Because it is not tested there, other people will think that everything is valid. Up until someone will try to use `green-mbpt` and fail, and only then we will see that there is a problem. To fix that problem, we would have to update `green-integrals` and then trigger CI on `green-mbpt` to detect that error was fixed.
 
 In my opinion that will create much more problems in the long run than potential extra book-keeping.

iskakoff commented on src/green/mbpt/except.h:
```diff
@@ -40,9 +40,5 @@ namespace green::mbpt {
     explicit mbpt_chemical_potential_search_failure(const std::string& what) : std::runtime_error(what) {}
   };
 
-  class mbpt_outdated_input : public std::runtime_error {
```
Same problem here, moving error class to `green-integrals` does not make sense, since the context of an error is different. I propose to bring it back.

iskakoff commented on src/green/mbpt/mbpt_run.h:
```diff
@@ -231,15 +231,17 @@ namespace green::mbpt {
       std::string int_version = ar.get_attribute<std::string>("__green_version__");
       if (int_version.rfind(INPUT_VERSION, 0) != 0) {
-        throw mbpt_outdated_input("Input file at '" + path +"' is outdated, please run migration script python/migrate.py");
+        throw integrals_outdated_input("Input file at '" + path +"' is outdated, please run migration script python/migrate.py");
```
Exactly the reason, `input.h5` is not `integrals`, this makes naming confusing.

iskakoff commented on src/green/mbpt/mbpt_run.h:
```diff
@@ -231,15 +231,17 @@ namespace green::mbpt {
       }
     } else {
-      throw mbpt_outdated_input("Input file at '" + path +"' is outdated, please run migration script python/migrate.py");
+      throw integrals_outdated_input("Input file at '" + path +"' is outdated, please run migration script python/migrate.py");
```
Same here

iskakoff commented on src/gf2_solver.cpp:
```diff
@@ -40,12 +40,16 @@ namespace green::mbpt {
+    // TODO: REVISIT! Possible source of out-of-memory segmentation faults
+    // There are several shared memory integral objects being formed here and by default,
+    // each integral occupies at-most 50% of memory (should be tunable). For large realistic jobs, this will cause problems.
+    int verbose_ints = (!utils::context.internode_rank) ? 1 : 0;
```
Why does only the first integral object have verbosity?

gauravharsha replied:
All the integral object are identical. Making all of them verbose doesn't make sense.

gauravharsha replied:
Also, something to be figure out: now that we have shared memory integrals and we are reading multiple chunks at the same time, perhaps the we could get rid of 6 integrals objects and just have 1.

iskakoff commented on src/gw_cpu_kernel.cpp:
```diff
@@ -21,15 +21,16 @@
 
   void gw_cpu_kernel::solve(G_type& g, St_type& sigma_tau) {
-    _coul_int1 = new df_integral_t(_path, _nao, _NQ, _bz_utils);
+    int verbose_ints = (!utils::context.internode_rank) ? 1 : 0;
```
In most cases you put `verbose_ints = (!utils::context.internode_rank) ? 1 : 0;`,
why don't we make the default value for `verbose = -1` in the `df_integral_t` declaration, and later in the definition use `(!utils::context.internode_rank)` condition if it's negative? Would make code cleaner.

gauravharsha replied:
On second thoughts, this and other similar segments should not make integrals verbose by default. Instead, the desired behavior is to switch "ON" the verbosity if the input parameter `--verbose` is set to be larger than 1. To make the code cleaner, I am defining `_verbose_ints` as a class variable, similar to `_verbose`.

iskakoff commented on src/hf_cpu_kernels.cpp:
```diff
@@ -27,7 +27,8 @@ namespace green::mbpt::kernels {
     {
-      df_integral_t coul_int1(_hf_path, _nao, _NQ, _bz_utils);
+      int verbose_ints = (!utils::context.internode_rank) ? 1 : 0;
+      df_integral_t coul_int1(_hf_path, _nao, _NQ, _bz_utils, verbose_ints);
```
Same `verbose_ints` here.

iskakoff commented on src/hf_cpu_kernels.cpp:
```diff
@@ -152,7 +144,8 @@ namespace green::mbpt::kernels {
     new_Fock.set_zero();
     {
-      df_integral_t coul_int1(_hf_path, _nao, _NQ, _bz_utils);
+      int verbose_ints = (!utils::context.internode_rank) ? 1 : 0;
```
And here (`verbose_ints`).

iskakoff commented on test/solvers_test.cpp:
```diff
@@ -199,6 +199,7 @@ TEST_CASE("MBPT Solver") {
   }
 
   SECTION("Input Data Version") {
+    SKIP("Input version test is disabled until all integral data and throw statement is updated");
```
Exactly because of that, I propose to keep `mbpt_outdated_input` to do `input.h5` version testing. And introduce separate version testing for integrals in `green-integrals`. To keep thing close to what they are needed for, importing exceptions from other repository if they do something close to what we need, but not exactly what we need can cause confusions, and also it breaks locality of tests.

iskakoff commented on test/solvers_test.cpp:
```diff
@@ -215,10 +216,10 @@ TEST_CASE("MBPT Solver") {
-    REQUIRE_THROWS_AS(green::mbpt::df_integral_t(df_int_path_2, 2, 36, bz_utils), green::mbpt::mbpt_outdated_input);
-    REQUIRE_NOTHROW(green::mbpt::df_integral_t(df_int_path_3, 2, 36, bz_utils));
-    REQUIRE_THROWS_AS(green::mbpt::check_input(p), green::mbpt::mbpt_outdated_input);
+    REQUIRE_THROWS_AS(green::mbpt::df_integral_t(df_int_path_1, 2, 36, bz_utils, p["verbose"]), green::integrals::integrals_outdated_input);
```
These three lines related to `df_integral_t` have to be moved to `green-integrals` repository, where they are defined.

iskakoff commented on test/test_main.cpp:
Also, what is this needed for?

gauravharsha replied:
Removed

iskakoff commented on LICENSE:
```diff
@@ -1,6 +1,4 @@
-MIT License
-
-Copyright (c) 2023 Green-Phys
+Copyright (c) 2024 Green-Phys
```
```suggestion
Copyright (c) 2023-2025 Green-Phys
```

gauravharsha commented:
@iskakoff The latest commit addresses all the comments mentioned in your review. I have also moved the tests for checking the version of integrals to [green-integrals](https://github.com/Green-Phys/green-integrals) (in [PR 6](https://github.com/Green-Phys/green-integrals/pull/6)).

gauravharsha commented:
Closing PR for now as we have moved away from buffered integrals for now
